llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -4585,6 +4585,50 @@ static Value *simplifySelectWithFakeICmpEq(Value *CmpLHS, Value *CmpRHS,
                                Pred == ICmpInst::ICMP_EQ);
 }
 
+/// Try to simplify the Select instruction consisting of and/or/xor.
+static Value *simplifySelectWithAndOrXorBitOp(Value *CmpLHS, Value *CmpRHS,
+                                              Value *TrueVal, Value *FalseVal) {
+  Value *X, *Y;
+  // (X & Y) == 0 ? X | Y : X ^ Y   --> X ^ Y
+  // (X & Y) == 0 ? X ^ Y : X | Y   --> X | Y
+  if (match(CmpLHS, m_And(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) {
+    if ((match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y))) &&
+         match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) ||
+        (match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y))) &&
+         match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))))
+      return FalseVal;
+  }
+  // (X | Y) == 0 ? X & Y : X ^ Y   --> X ^ Y
+  // (X | Y) == 0 ? X ^ Y : X & Y   --> X & Y
+  if (match(CmpLHS, m_Or(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) {
+    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) ||
+        (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))))
+      return FalseVal;
+  }
+  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
+  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
+  if (match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) {
```

Comment:
Does `X ^ Y == 0` ever not get canonicalized to `X == Y`?
If so I think this case is dead and you can just rely on the the below one (although you should keep the comments).

Reply:
When running with instcombine, xor is transformed to eq and comes. In case of instsimplify, it comes as xor.

If xor comes, is it desirable to process it by transform it to eq?

Reply:
What you could also do to minimize the code duplication is do the following:

```
if ((match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) || (match(CmpLHS, m_Value(X)) && match(CmpRhs, m_Value(Y)))) {
```

Then drop the lines below.

@nikic, is there a general policy for whether we can rely on `InstCombine` canonicalizations in `InstSimplify`?

Reply:
InstSimplify should generally assume canonical form, unless there is some specific reason to the contrary.

Reply:
sure. however, I would like you to confirm that my understanding is correct.

"Change that remove X^Y transforming, leaving only X==Y, but retaining the comment." 

correct?

Reply:
@goldsteinn like this?
```
  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
  // X == Y is equals with (X ^ Y) == 0
  LLVM_DEBUG(dbgs() << "xor\n");
  if ((match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) &&
       match(CmpRHS, m_Zero())) ||
      (match(CmpLHS, m_Value(X) && match(CmpRhs, m_Value(Y))))) {
    LLVM_DEBUG(dbgs() << "xor inside\n");
    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
         match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y)))) ||
        (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
         match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))))
      return FalseVal;
  }
```

Reply:
No need for the explciit xor == 0 match. just:

```
  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
 // X == Y is equals with (X ^ Y) == 0
  X = CmpLHS;
  Y = CmpRHS;
  if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
       match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y)))) ||
      (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
       match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))))
    return FalseVal;
```

is fine.